book-content/chapters/05-unions-literals-and-narrowing.md

```diff
@@ -0,0 +1,1856 @@
+  // Move the specified distance in the given direction
+}
+```
+
+The implementation of the function is not important, but the idea is that we want to be able to move either up, down, left, or right.
+
+Here's what calling the `move` function might look like:
+
+```typescript
+move("up", 10);
+
+move("left", 5);
+```
+
+To test this function, we have some `@ts-expect-error` directives that tell TypeScript we expect the following lines to throw an error.
+
+However, since the `move` function currently takes in a `string` for the `direction` parameter, we can pass in any string we want, even if it's not a valid direction. There is also a test where we expect that passing `20` as a distance won't work, but it's being accepted as well.
+
+This leads to TypeScript drawing red squiggly lines under the `@ts-expect-error` directives:
+
+```ts twoslash
+// @errors: 2578
+function move(direction: string, distance: number) {
+  // Move the specified distance in the given direction
+}
+// ---cut---
+move(
+  // @ts-expect-error - "up-right" is not a valid direction
+  "up-right",
+  10,
+);
+
+move(
+  // @ts-expect-error - "down-left" is not a valid direction
+  "down-left",
+  20,
+);
+```
+
+Your challenge is to update the `move` function so that it only accepts the strings `"up"`, `"down"`, `"left"`, and `"right"`. This way, TypeScript will throw an error when we try to pass in any other string.
+
+<Exercise title="Exercise 2: Restricting Function Parameters" filePath="/src/018-unions-and-narrowing/054-literal-types.problem.ts"></Exercise>
+
+#### Solution 1: `string` or `null`
+
+The solution is to update the `username` parameter to be a union of `string` and `null`:
+
+```typescript
+function getUsername(username: string | null) {
+  // function body
+}
+```
+
+With this change, the `getUsername` function will now accept `null` as a valid value for the `username` parameter, and the errors will be resolved.
+
+#### Solution 2: Restricting Function Parameters
+
+In order to restrict what the `direction` can be, we can use a union type of literal values (in this case strings).
+
+Here's what this looks like:
+
+```typescript
+function move(direction: "up" | "down" | "left" | "right", distance: number) {
+  // Move the specified distance in the given direction
+}
+```
+
+With this change, we now have autocomplete for the possible `direction` values.
+
+To clean things up a bit, we can create a new type alias called `Direction` and update the parameter accordingly:
+
+```typescript
+type Direction = "up" | "down" | "left" | "right";
+
+function move(direction: Direction, distance: number) {
+  // Move the specified distance in the given direction
+}
+```
+
+## Narrowing
+
+### Wider vs Narrower Types
+
+Some types are wider versions of other types. For example, `string` is wider than the literal string `"small"`. This is because `string` can be any string, while `"small"` can only be the string `"small"`.
```

Comment:
```suggestion
Some types are wider versions of other types. For example, `string` is wider than the literal string `"small"`. This is because `string` can be any string (including `"small"`), while `"small"` can only be the string `"small"`.
```

Reply:
but i guess u get to that in a sec, so there might not be a need to mention it straightaway